chains/evm/listener/event-handler.go

```diff
@@ -101,50 +100,24 @@ func toCallArg(msg ethereum.CallMsg) map[string]interface{} {
 	return arg
 }
 
-func Erc20EventHandler(sourceID, destId uint8, nonce uint64, handlerContractAddress common.Address, client ChainClient) (*relayer.Message, error) {
-	definition := "[{\"inputs\":[{\"internalType\":\"uint64\",\"name\":\"depositNonce\",\"type\":\"uint64\"},{\"internalType\":\"uint8\",\"name\":\"destId\",\"type\":\"uint8\"}],\"name\":\"getDepositRecord\",\"outputs\":[{\"components\":[{\"internalType\":\"address\",\"name\":\"_tokenAddress\",\"type\":\"address\"},{\"internalType\":\"uint8\",\"name\":\"_lenDestinationRecipientAddress\",\"type\":\"uint8\"},{\"internalType\":\"uint8\",\"name\":\"_destinationDomainID\",\"type\":\"uint8\"},{\"internalType\":\"bytes32\",\"name\":\"_resourceID\",\"type\":\"bytes32\"},{\"internalType\":\"bytes\",\"name\":\"_destinationRecipientAddress\",\"type\":\"bytes\"},{\"internalType\":\"address\",\"name\":\"_depositer\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"_amount\",\"type\":\"uint256\"}],\"internalType\":\"structERC20Handler.DepositRecord\",\"name\":\"\",\"type\":\"tuple\"}],\"stateMutability\":\"view\",\"type\":\"function\"}]"
-	type ERC20HandlerDepositRecord struct {
-		TokenAddress                   common.Address
-		LenDestinationRecipientAddress uint8
-		DestinationDomainID            uint8
-		ResourceID                     [32]byte
-		DestinationRecipientAddress    []byte
-		Depositer                      common.Address
-		Amount                         *big.Int
-	}
-	a, err := abi.JSON(strings.NewReader(definition))
-	if err != nil {
-		return nil, err
-	}
-
-	input, err := a.Pack("getDepositRecord", nonce, destId)
-	if err != nil {
+// Erc20EventHandler converts data pulled from event logs into message
+// handlerResponse can be an empty slice
+func Erc20EventHandler(sourceID, destId uint8, nonce uint64, resourceID [32]byte, addr common.Address, calldata, handlerResponse []byte) (*relayer.Message, error) {
+	if len(calldata) == 0 {
+		err := errors.New("missing calldata")
 		return nil, err
 	}
 
-	msg := ethereum.CallMsg{From: common.Address{}, To: &handlerContractAddress, Data: input}
-	out, err := client.CallContract(context.TODO(), toCallArg(msg), nil)
-	if err != nil {
-		return nil, err
-	}
-	res, err := a.Unpack("getDepositRecord", out)
-	if err != nil {
-		return nil, err
-	}
-	if len(res) == 0 {
-		return nil, errors.New("no handler associated with such resourceID")
-	}
-
-	out0 := *abi.ConvertType(res[0], new(ERC20HandlerDepositRecord)).(*ERC20HandlerDepositRecord)
 	return &relayer.Message{
 		Source:       sourceID,
 		Destination:  destId,
 		DepositNonce: nonce,
-		ResourceId:   out0.ResourceID,
+		ResourceId:   resourceID,
 		Type:         relayer.FungibleTransfer,
 		Payload: []interface{}{
-			out0.Amount.Bytes(),
-			out0.DestinationRecipientAddress,
+			calldata,
```

Comment:
It should not work this way. Because check what ERC20MessageHandler expects to see in Payload, ther is 2 way, you can parse calldata here and pass expected data for messageHnalder or change how the handler behaves, id rather select second

Reply:
done